File: DDRolLocal/fallback/python/src/nyx_fallback/core/knowledge_acquisition.py
"""Admission of new acquisitions only. Historical replay deliberately does not call this policy."""

from typing import Any

from .canonical import canonical_json
from .contracts import validate_contract


def _at(value: Any, key: Any) -> Any:
    return value.get(key) if isinstance(value, dict) and isinstance(key, str) else None


def _text(value: Any) -> str | None:
    text = _at(value, "text")
    return text if isinstance(text, str) and text else None


def validate_new_knowledge_events(state: dict[str, Any], events: list[dict[str, Any]], narration: str, player_input: str) -> None:
    learning = [event for event in events if _at(event, "event_type") == "NPC_KNOWLEDGE_ADDED"]
    if not learning:
        return
    if len(learning) != len(events):
        raise ValueError("KNOWLEDGE_MIXED_TRANSACTION")
    present = state["scene"].get("present_character_ids", [])
    if not isinstance(present, list):
        present = []
    receipts: list[str] = []
    learned: set[str] = set()
    for event in learning:
        payload = event.get("payload")
        if (not isinstance(payload, dict) or set(payload) != {"acquisition", "subject_id", "knowledge_id", "value"}
                or validate_contract("internal/knowledge-acquisition-v1.schema.json", payload["acquisition"])):
            raise ValueError("KNOWLEDGE_PROVENANCE_REQUIRED")
        proof = payload["acquisition"]
        subject = payload["subject_id"]
        receiver = _at(state["characters"], subject)
        receiver_name = _at(_at(receiver, "sheet"), "name")
        if not isinstance(receiver, dict) or subject not in present or not isinstance(receiver_name, str) or not receiver_name:
            raise ValueError("KNOWLEDGE_SOURCE_UNCONFIRMED")
        receiver_knowledge = _at(state["knowledge"], subject)
        identity = f"{subject}:{canonical_json(payload['value'])}"
        if (not isinstance(payload["knowledge_id"], str) or identity in learned
                or (isinstance(receiver_knowledge, dict) and (payload["knowledge_id"] in receiver_knowledge
                    or any(canonical_json(value) == canonical_json(payload["value"]) for value in receiver_knowledge.values())))):
            raise ValueError("KNOWLEDGE_ALREADY_ACQUIRED")
        learned.add(identity)
        if proof["kind"] == "DISCLOSURE":
            source_id = proof["source_subject_id"]
            source = _at(state["characters"], source_id)
            source_name = _at(_at(source, "sheet"), "name")
            expected = _at(_at(state["knowledge"], source_id), proof["source_id"])
            text = _text(expected)
            if (not isinstance(source, dict) or not isinstance(source_name, str) or not source_name or not text
                    or source_id not in present or source_id == subject or event.get("actor_id") != source_id):
                raise ValueError("KNOWLEDGE_SOURCE_UNCONFIRMED")
            if isinstance(expected, dict) and "secret_id" in expected:
                secret = _at(state["canon"]["secrets"], expected["secret_id"])
                audience = _at(secret, "authorized_subject_ids")
                readers = [character["character_id"] for character in state["characters"].values() if character["kind"] == "player"]
                if not isinstance(audience, list) or not all(member in audience for member in [subject, source_id, *readers]):
                    raise ValueError("KNOWLEDGE_AUDIENCE_BLOCKED")
            reference = f"knowledge:{source_id}:{proof['source_id']}"
            receipt = f"{source_name} dice en privado a {receiver_name}: «{text}»"
            if source["kind"] == "player" and player_input != receipt:
                raise ValueError("KNOWLEDGE_PLAYER_AGENCY_BLOCKED")
        else:
            fact = _at(state["canon"]["facts"], proof["source_id"])
            text = _text(fact)
            if (proof["source_subject_id"] is not None or not isinstance(fact, dict) or not text
                    or fact.get("visibility") != "public" or not isinstance(fact.get("location_id"), str)
                    or fact["location_id"] != state["scene"].get("location_id")
                    or fact["location_id"] != state["world"].get("location_id") or event.get("actor_id") != subject):
                raise ValueError("KNOWLEDGE_SOURCE_UNCONFIRMED")
            expected = {"fact_id": proof["source_id"], "text": text}
            reference = f"canon:{proof['source_id']}"
            receipt = f"{receiver_name} observa: «{text}»"
        if canonical_json(payload["value"]) != canonical_json(expected):
            raise ValueError("KNOWLEDGE_CLAIM_MISMATCH")
        if not all(isinstance(event.get(key), list) and needle in event[key]
                   for key, needle in (("source_refs", reference), ("evidence", reference), ("targets", subject))):
            raise ValueError("KNOWLEDGE_SOURCE_UNCONFIRMED")
        receipts.append(receipt)
    if narration != " ".join(receipts):
        raise ValueError("KNOWLEDGE_NARRATION_MISMATCH")

File: DDRolLocal/packages/domain/src/campaign/knowledge-acquisition.ts
import { validateRuntimeContract, type InternalKnowledgeAcquisitionV1 } from '@nyx/contracts';
import { canonicalJson } from '../canonical-json.js';
import { err, ok, type Result } from '../result.js';
import type { CampaignState } from '../state/types.js';

type Proof = InternalKnowledgeAcquisitionV1.KnowledgeAcquisitionV1;
const fail = (code: string) => err({ code, message: `Adquisición de conocimiento bloqueada: ${code}.` });
function record(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}
function own(value: unknown, key: unknown): unknown {
  return record(value) && typeof key === 'string' && Object.hasOwn(value, key) ? value[key] : undefined;
}
function textOf(value: unknown): string | undefined {
  const text = own(value, 'text');
  return typeof text === 'string' && text.length > 0 ? text : undefined;
}

/** New event admission only, never replay/migration. Sources are resolved against the confirmed
 * pre-transaction snapshot: a proposal cannot manufacture a source earlier in the same batch.
 * Exact receipts bind acquisition to visible communication, without an NLP entailment engine.
 */
export function validateNewKnowledgeEvents(state: CampaignState, events: readonly unknown[], narration: string,
  playerInput: string): Result<true, { code: string; message: string }> {
  const learning = events.filter(event => own(event, 'event_type') === 'NPC_KNOWLEDGE_ADDED');
  if (learning.length === 0) return ok(true);
  if (learning.length !== events.length) return fail('KNOWLEDGE_MIXED_TRANSACTION');
  const present = Array.isArray(state.scene['present_character_ids']) ? state.scene['present_character_ids'] : [];
  const receipts: string[] = [];
  const learned = new Set<string>();
  for (const candidate of learning) {
    if (!record(candidate) || !record(candidate['payload'])) return fail('KNOWLEDGE_PROVENANCE_REQUIRED');
    const p = candidate['payload'];
    if (Object.keys(p).sort().join(',') !== 'acquisition,knowledge_id,subject_id,value'
      || !validateRuntimeContract('KnowledgeAcquisition', p['acquisition']).ok) return fail('KNOWLEDGE_PROVENANCE_REQUIRED');
    const proof = p['acquisition'] as Proof;
    const subject = p['subject_id'];
    const receiver = own(state.characters, subject);
    const receiverName = own(own(receiver, 'sheet'), 'name');
    if (typeof subject !== 'string' || !record(receiver) || !present.includes(subject)
      || typeof receiverName !== 'string' || !receiverName.length) return fail('KNOWLEDGE_SOURCE_UNCONFIRMED');
    const receiverKnowledge = own(state.knowledge, subject);
    const identity = `${subject}:${canonicalJson(p['value'])}`;
    if (typeof p['knowledge_id'] !== 'string' || own(receiverKnowledge, p['knowledge_id']) !== undefined || learned.has(identity)
      || (record(receiverKnowledge) && Object.values(receiverKnowledge).some(value => canonicalJson(value) === canonicalJson(p['value'])))) {
      return fail('KNOWLEDGE_ALREADY_ACQUIRED');
    }
    learned.add(identity);
    let expected: unknown;
    let text: string | undefined;
    let reference: string;
    let receipt: string;
    if (proof.kind === 'DISCLOSURE') {
      const source = own(state.characters, proof.source_subject_id);
      const sourceName = own(own(source, 'sheet'), 'name');
      expected = own(own(state.knowledge, proof.source_subject_id), proof.source_id);
      text = textOf(expected);
      if (!record(source) || typeof sourceName !== 'string' || !sourceName.length || !text
        || !present.includes(proof.source_subject_id) || proof.source_subject_id === subject
        || candidate['actor_id'] !== proof.source_subject_id) return fail('KNOWLEDGE_SOURCE_UNCONFIRMED');
      const secretId = own(expected, 'secret_id');
      if (secretId !== undefined) {
        const secret = own(state.canon.secrets, secretId);
        const audience = own(secret, 'authorized_subject_ids');
        const readers = Object.values(state.characters).filter(character => character.kind === 'player').map(character => character.character_id);
        if (!Array.isArray(audience) || ![subject, proof.source_subject_id, ...readers].every(id => audience.includes(id))) {
          return fail('KNOWLEDGE_AUDIENCE_BLOCKED');
        }
      }
      reference = `knowledge:${proof.source_subject_id}:${proof.source_id}`;
      receipt = `${sourceName} dice en privado a ${receiverName}: «${text}»`;
      if (own(source, 'kind') === 'player' && playerInput !== receipt) return fail('KNOWLEDGE_PLAYER_AGENCY_BLOCKED');
    } else {
      const fact = own(state.canon.facts, proof.source_id);
      text = textOf(fact);
      if (proof.source_subject_id !== null || !record(fact) || !text || fact['visibility'] !== 'public'
        || typeof fact['location_id'] !== 'string' || fact['location_id'] !== state.scene['location_id']
        || fact['location_id'] !== state.world['location_id'] || candidate['actor_id'] !== subject) return fail('KNOWLEDGE_SOURCE_UNCONFIRMED');
      expected = { fact_id: proof.source_id, text };
      reference = `canon:${proof.source_id}`;
      receipt = `${receiverName} observa: «${text}»`;
    }
    if (canonicalJson(p['value']) !== canonicalJson(expected)) return fail('KNOWLEDGE_CLAIM_MISMATCH');
    if (!Array.isArray(candidate['source_refs']) || !candidate['source_refs'].includes(reference)
      || !Array.isArray(candidate['evidence']) || !candidate['evidence'].includes(reference)
      || !Array.isArray(candidate['targets']) || !candidate['targets'].includes(subject)) return fail('KNOWLEDGE_SOURCE_UNCONFIRMED');
    receipts.push(receipt);
  }
  return narration === receipts.join(' ') ? ok(true) : fail('KNOWLEDGE_NARRATION_MISMATCH');
}
